tb_plugin/torch_tb_profiler/profiler/data.py

```diff
@@ -138,3 +175,83 @@ def analyze(self):
+                "#fuse-pointwise-operations",
+                "https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html"
+                "#checkpoint-intermediate-buffers",
+                "https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html"
+                "#avoid-unnecessary-cpu-gpu-synchronization",
+                "https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html"
+                "#create-tensors-directly-on-the-target-device"
+            )
+            self.recommendations.append(text)
+
+        if self.runtime_node_list is not None and len(self.runtime_node_list) > 0:
+            total_kernels = 0
+            short_kernels = 0
+            for rt in self.runtime_node_list:
+                if rt.device_nodes is not None:
+                    for node in rt.device_nodes:
+                        if node.type == EventTypes.KERNEL:
+                            total_kernels += 1
+                            if node.end_time - node.start_time < rt.end_time - rt.start_time:
+                                short_kernels += 1
+            if short_kernels / total_kernels > 0.5 and total_kernels > 100:
+                text = "{} out of {} kernels are short in execution time. " \
+                       "You could try to <a href =\"{}\" target=\"_blank\">optimize zero_grad</a>, " \
+                       "or <a href =\"{}\" target=\"_blank\">fuse pointwise operations</a>.".format(
+                    short_kernels, total_kernels,
+                    "https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html"
+                    "#use-parameter-grad-none-instead-of-model-zero-grad-or-optimizer-zero-grad",
+                    "https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html"
+                    "#fuse-pointwise-operations"
+                )
+            self.recommendations.append(text)
+
+        low_sm_efficiency_gpus = []
+        for gpu_id in self.gpu_ids:
+            if self.sm_efficency[gpu_id] > 0 and self.sm_efficency[gpu_id] < 0.8 * self.gpu_utilization[gpu_id]:
+                low_sm_efficiency_gpus.append(gpu_id)
+        if len(low_sm_efficiency_gpus) > 0:
+            gpu_list_str, has_str = get_gpus_str(low_sm_efficiency_gpus)
+            text = "GPU {} {} low estimated SM efficiency. " \
+                   "Many kernels' launched blocks are too few that they can't fully utilize all multiprocessors." \
+                   "You could try to increase the blocks number of these kernels.".format(
+                gpu_list_str, has_str)
+            self.recommendations.append(text)
+
+        low_occupancy_gpus = []
+        for gpu_id in self.gpu_ids:
+            if self.occupancy[gpu_id] > 0 and self.occupancy[gpu_id] < 50:
+                low_occupancy_gpus.append(gpu_id)
+        if len(low_occupancy_gpus) > 0:
+            gpu_list_str, has_str = get_gpus_str(low_occupancy_gpus)
```

Comment:
suggest to move the gpu related metrics to a separated file and only use the output here.

Reply:
This is the rule based recommendation. We can refactor the recommendation to another file, but maybe not in this PR.

Reply:
could these logic to be moved into a separated method instead?

Reply:
Thanks. Done.